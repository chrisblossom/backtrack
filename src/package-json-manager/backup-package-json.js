/* @flow */

import path from 'path';
import { get, isEqual } from 'lodash';
import { backupFile } from '../utils/backup-file';
import { rootPath } from '../config/paths';
import log from '../utils/log';
import { mapObjectKeyNames } from '../utils/object-utils';

import type { PackageJson } from '../types.js';

type Return = Promise<?{
    file: string,
    filename: string,
}>;

const npmDefaults = {
    version: '1.0.0',
    description: '',
    main: 'index.js',
    scripts: {
        test: 'echo "Error: no test specified" && exit 1',
    },
    keywords: [],
    author: '',
    license: 'ISC',
};

async function backupPackageJson(
    packageJson: PackageJson,
    managedKeys: ?PackageJson,
    previousManagedKeys: ?PackageJson,
): Return {
    const mapManagedKeys = mapObjectKeyNames(managedKeys);
    /**
     * backup package.json if any custom managed scripts exist
     */
    const shouldBackup = mapManagedKeys.some((key) => {
        const matchedMangedKey = get(managedKeys, key);
        const matchedPackageJson = get(packageJson, key);
        const matchedPreviousManagedKey = get(previousManagedKeys, key);
        const matchedDefaultKey = get(npmDefaults, key);

        /**
         * Don't backup if package.json key === npm's default
         */
        if (
            matchedPreviousManagedKey === undefined &&
            isEqual(matchedPackageJson, matchedDefaultKey)
        ) {
            return false;
        }

        // use isEqual to match arrays
        const result =
            matchedPackageJson &&
            isEqual(matchedMangedKey, matchedPackageJson) === false &&
            isEqual(matchedPreviousManagedKey, matchedPackageJson) === false;

        return result;
    });

    if (shouldBackup === false) {
        return undefined;
    }

    const packageJsonFile = path.resolve(rootPath, 'package.json');
    const backupData = await backupFile(packageJsonFile);

    if (!backupData) {
        return undefined;
    }

    const relativeBackupFilePath = path.relative(rootPath, backupData.file);

    log.warn(
        `Unknown package.json scripts. Backing up to: ${relativeBackupFilePath}`,
    );

    return {
        file: backupData.file,
        filename: backupData.filename,
    };
}

export { backupPackageJson };
