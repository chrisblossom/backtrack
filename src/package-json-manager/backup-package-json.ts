import path from 'path';
import { get, isEqual } from 'lodash';
import { backupFile } from '../utils/backup-file';
import { rootPath } from '../config/paths';
import log from '../utils/log';
import { mapObjectKeyNames } from '../utils/object-utils';

import { PackageJson } from '../types';

type Return = Promise<{
    file: string;
    filename: string;
} | void>;

const npmDefaults = {
    version: '1.0.0',
    description: '',
    main: 'index.js',
    scripts: {
        test: 'echo "Error: no test specified" && exit 1',
    },
    keywords: [],
    author: '',
    license: 'ISC',
};

async function backupPackageJson(
    packageJson: PackageJson,
    managedKeys?: PackageJson,
    previousManagedKeys?: PackageJson,
): Return {
    const mapManagedKeys = mapObjectKeyNames(managedKeys);
    /**
     * backup package.json if any custom managed scripts exist
     */

    const shouldBackup = mapManagedKeys
        .filter((key) => {
            const matchedMangedKey = get(managedKeys, key);
            const matchedPackageJson = get(packageJson, key);
            const matchedPreviousManagedKey = get(previousManagedKeys, key);
            const matchedDefaultKey = get(npmDefaults, key);

            /**
             * Don't backup if package.json key === npm default
             */
            if (
                matchedPreviousManagedKey === undefined &&
                isEqual(matchedPackageJson, matchedDefaultKey)
            ) {
                return false;
            }

            // use isEqual to match arrays
            const result =
                matchedPackageJson &&
                isEqual(matchedMangedKey, matchedPackageJson) === false &&
                isEqual(matchedPreviousManagedKey, matchedPackageJson) ===
                    false;

            return result;
        })
        .map((backupKeys) => {
            return backupKeys.join('.');
        });

    if (shouldBackup.length === 0) {
        return undefined;
    }

    const packageJsonFile = path.resolve(rootPath, 'package.json');
    const backupData = await backupFile(packageJsonFile);

    if (!backupData) {
        return undefined;
    }

    const relativeBackupFilePath = path.relative(rootPath, backupData.file);

    log.warn(
        `Unknown package.json keys: "${shouldBackup.join(
            ', ',
        )}". Backing up to: ${relativeBackupFilePath}`,
    );

    return {
        file: backupData.file,
        filename: backupData.filename,
    };
}

export { backupPackageJson };
