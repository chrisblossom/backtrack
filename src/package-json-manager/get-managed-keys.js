/* @flow */

import { merge } from 'lodash';

import type { Lifecycles, PackageJson } from '../types.js';

const customScripts = {
    dev: 'backtrack dev --development',
    build: 'backtrack build --production',

    // 'dev.prod': 'backtrack dev --production',
    // 'build.dev': 'backtrack build --development',
};

function getManagedKeys(lifecycles: Lifecycles = {}): PackageJson {
    const { packageJson = [] } = lifecycles;

    const getScripts = Object.keys(lifecycles).reduce(
        (acc, currentLifecycle) => {
            /**
             * Remove internally managed tasks
             */
            if (
                ['packageJson', 'config', 'files', 'resolve'].includes(
                    currentLifecycle,
                )
            ) {
                return acc;
            }

            const script = lifecycles[currentLifecycle];

            /**
             * Remove unused lifecycles
             */
            if (!script || (Array.isArray(script) && script.length === 0)) {
                return acc;
            }

            /**
             * call backtrack LIFECYCLE unless modified
             */
            const matched =
                // $FlowIssue
                customScripts[currentLifecycle] ||
                `backtrack ${currentLifecycle}`;

            return { ...acc, [currentLifecycle]: matched };
        },
        {},
    );

    const scripts =
        Object.keys(getScripts).length !== 0 ? { scripts: getScripts } : {};

    const mergedKeys: PackageJson = merge(scripts, ...packageJson);

    return mergedKeys;
}

export { getManagedKeys };
