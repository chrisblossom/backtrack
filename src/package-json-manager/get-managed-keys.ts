import { merge } from 'lodash';
import { plainObjectHasKey, Lifecycles, PackageJson } from '../types';

const customScripts = {
	dev: 'backtrack dev --development',
	build: 'backtrack build --production',

	// 'dev.prod': 'backtrack dev --production',
	// 'build.dev': 'backtrack build --development',
};

function getManagedKeys(lifecycles: Lifecycles = {}): PackageJson {
	const { packageJson = [] } = lifecycles;

	const lifeCycleKeys = Object.keys(lifecycles);
	const getScripts = lifeCycleKeys.reduce((acc, currentLifecycle) => {
		/**
		 * Remove internally managed tasks
		 */
		if (
			[
				'packageJson',
				'config',
				'files',
				'resolve',
			].includes(currentLifecycle)
		) {
			return acc;
		}

		const script = lifecycles[currentLifecycle];

		/**
		 * Remove unused lifecycles
		 */
		if (!script || (Array.isArray(script) && script.length === 0)) {
			return acc;
		}

		/**
		 * call backtrack LIFECYCLE unless modified
		 */
		if (plainObjectHasKey(customScripts, currentLifecycle)) {
			const matched = customScripts[currentLifecycle];

			return {
				...acc,
				[currentLifecycle]: matched,
			};
		}

		const matched = `backtrack ${currentLifecycle}`;
		return { ...acc, [currentLifecycle]: matched };
	}, {});

	const scripts =
		Object.keys(getScripts).length !== 0 ? { scripts: getScripts } : {};

	const mergePackageJson = merge({}, ...packageJson);
	const mergedKeys: PackageJson = merge(scripts, mergePackageJson);

	return mergedKeys;
}

export { getManagedKeys };
