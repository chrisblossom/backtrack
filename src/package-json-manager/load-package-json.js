/* @flow */

import path from 'path';
import { existsSync } from 'fs';
import { rootPath } from '../config/paths';

import type { PackageJson } from '../types.js';

function loadPackageJson(): PackageJson {
    const file = path.resolve(rootPath, 'package.json');
    const exists = existsSync(file);

    if (exists === false) {
        throw new Error('package.json not found');
    }

    const load = require(file);
    return load;
}

export { loadPackageJson };
