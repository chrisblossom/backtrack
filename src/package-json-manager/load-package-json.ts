import path from 'path';
import { existsSync } from 'fs';
import { rootPath } from '../config/paths';

import { PackageJson } from '../types';

function loadPackageJson(): PackageJson {
    const file = path.resolve(rootPath, 'package.json');
    const exists = existsSync(file);

    if (exists === false) {
        throw new Error('package.json not found');
    }

    const load = require(file);
    return load;
}

export { loadPackageJson };
