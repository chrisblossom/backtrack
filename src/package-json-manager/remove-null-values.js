/* @flow */

import { isPlainObject } from 'lodash';
import type { PackageJson } from '../types';

function removeNullValues(
    packageJson: {} /* flow errors with PackageJson */,
): PackageJson {
    const removeNull = (object) => {
        return Object.keys(object).reduce((acc, key) => {
            let value = object[key];

            if (value === null) {
                return acc;
            }

            if (isPlainObject(value)) {
                value = removeNull(value);
            }

            return {
                ...acc,
                [key]: value,
            };
        }, {});
    };

    const result = removeNull(packageJson);

    return result;
}

export { removeNullValues };
