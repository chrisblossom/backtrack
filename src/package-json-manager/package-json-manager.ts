import { getManagedKeys } from './get-managed-keys';
import { shouldUpdate } from './should-update';
import { updatePackageJson } from './update-package-json';
import { backupPackageJson } from './backup-package-json';
import { loadPackageJson } from './load-package-json';
import { removeNullValues } from './remove-null-values';
import { sortPackageJson } from './sort-package-json';
import { writePackageJson } from './write-package-json';

import { Lifecycles, PackageJson } from '../types';

async function packageJsonManager(
	lifecycles: Lifecycles = {},
	previousManagedKeys?: PackageJson,
): Promise<PackageJson> {
	/**
	 * Get managed keys
	 */
	const managedKeys = getManagedKeys(lifecycles);

	/**
	 * load current package.json from file
	 */
	const packageJson = loadPackageJson();

	/**
	 * Check if update is needed
	 */
	const packageJsonIsOutdated = shouldUpdate(
		packageJson,
		managedKeys,
		previousManagedKeys,
	);
	if (packageJsonIsOutdated === false) {
		return managedKeys;
	}

	/**
	 * Check if package.json needs to be backed up
	 */
	await backupPackageJson(packageJson, managedKeys, previousManagedKeys);

	/**
	 * Merge managed keys with current packageJson
	 */
	const updatedPackageJson = updatePackageJson(
		packageJson,
		managedKeys,
		previousManagedKeys,
	);

	/**
	 * remove null values
	 */
	const packageJsonFiltered = removeNullValues(updatedPackageJson);

	/**
	 * sort package json
	 */
	const packageJsonSorted = sortPackageJson(packageJsonFiltered);

	/**
	 * write packageJson
	 */
	writePackageJson(packageJsonSorted);

	return managedKeys;
}

export { packageJsonManager };
