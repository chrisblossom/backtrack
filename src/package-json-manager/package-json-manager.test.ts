import path from 'path';
import fs from 'fs';

const packageJsonManager = (presets?: any, previousManagedKeys?: any) =>
    require('./package-json-manager').packageJsonManager(
        presets,
        previousManagedKeys,
    );

describe('packageJsonManager', () => {
    const cwd = process.cwd();
    let writeFileSync: any;

    beforeEach(() => {
        jest.mock('../utils/log', () => ({
            warn: jest.fn(),
            info: jest.fn(),
            error: jest.fn(),
            success: jest.fn(),
        }));

        jest.mock('fs-extra', () => ({
            move: jest.fn(() => Promise.resolve()),
        }));

        /**
         * Before calling '.toMatchSnapshot();' snapshot, call 'writeFileSync.mockRestore();'
         * otherwise snapshot will not write
         */
        writeFileSync = jest
            .spyOn(fs, 'writeFileSync')
            .mockImplementation(() => jest.fn());
    });

    afterEach(() => {
        process.chdir(cwd);
    });

    test('handles undefined', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/package-json-1/');
        process.chdir(dir);

        const result = await packageJsonManager();

        const writeFileSyncCalls = writeFileSync.mock.calls;

        writeFileSync.mockRestore();

        expect(writeFileSyncCalls).toEqual([]);
        expect(result).toEqual({});
    });

    test('adds custom scripts when they do not exist prior', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/package-json-1/');
        process.chdir(dir);

        const lifecycles = {
            packageJson: [
                {
                    scripts: {
                        'lint.fix': 'eslint --fix',
                    },
                },
            ],
        };

        const result = await packageJsonManager(lifecycles);

        const writeFileSyncCalls = writeFileSync.mock.calls;

        writeFileSync.mockRestore();

        expect(writeFileSyncCalls).toMatchSnapshot();
        expect(result).toEqual({ scripts: { 'lint.fix': 'eslint --fix' } });
    });

    test('removes empty npm defaults', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/package-json-1/');
        process.chdir(dir);

        const lifecycles = {
            packageJson: [
                {
                    license: '',
                    scripts: {
                        'lint.fix': 'eslint --fix',
                        test: null,
                    },
                },
            ],
        };

        const result = await packageJsonManager(lifecycles);

        const writeFileSyncCalls = writeFileSync.mock.calls;

        writeFileSync.mockRestore();

        expect(writeFileSyncCalls).toMatchSnapshot();
        expect(result).toEqual({
            license: '',
            scripts: {
                test: null,
                'lint.fix': 'eslint --fix',
            },
        });
    });

    test('setting null does not always update packageJson if already removed', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/package-json-6/');
        process.chdir(dir);

        const lifecycles = {
            packageJson: [
                {
                    scripts: {
                        test: null,
                    },
                },
            ],
        };

        const result = await packageJsonManager(lifecycles);

        const writeFileSyncCalls = writeFileSync.mock.calls;

        writeFileSync.mockRestore();

        expect(writeFileSyncCalls).toMatchSnapshot();
        expect(result).toEqual({ scripts: { test: null } });
    });
});
