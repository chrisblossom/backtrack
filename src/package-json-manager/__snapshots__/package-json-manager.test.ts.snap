// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packageJsonManager adds custom scripts when they do not exist prior 1`] = `
[
  [
    "<PROJECT_ROOT>/package.json",
    "{
  "version": "1.0.0",
  "description": "",
  "keywords": [],
  "license": "ISC",
  "private": true,
  "author": "",
  "main": "index.js",
  "scripts": {
    "lint.fix": "eslint --fix",
    "test": "echo /"Error: no test specified/" && exit 1"
  },
  "devDependencies": {},
  "dependencies": {}
}
",
  ],
]
`;

exports[`packageJsonManager removes empty npm defaults 1`] = `
[
  [
    "<PROJECT_ROOT>/package.json",
    "{
  "version": "1.0.0",
  "description": "",
  "keywords": [],
  "license": "",
  "private": true,
  "author": "",
  "main": "index.js",
  "scripts": {
    "lint.fix": "eslint --fix"
  },
  "devDependencies": {},
  "dependencies": {}
}
",
  ],
]
`;

exports[`packageJsonManager setting null does not always update packageJson if already removed 1`] = `[]`;
